src/runtime/vm/vm.cc

```diff
@@ -42,6 +44,51 @@ namespace tvm {
 namespace runtime {
 namespace vm {
 
+// Cutomized reclamation of buffers.
+class StorageAllocator :
+      public ObjAllocatorBase<StorageAllocator> {
+ public:
+  template<typename T>
+  class Handler {
+   public:
+    template<typename... Args>
+    static T* New(StorageAllocator*, Args&&... args) {
+      static_assert(std::is_base_of<StorageObj, T>::value,
+                  "make_storage can only be used to create storage objects");
+      // NOTE: the first argument is not needed for SimpleObjAllocator
+      // It is reserved for special allocators that needs to recycle
+      // the object to itself (e.g. in the case of object pool).
+      //
+      // In the case of an object pool, an allocator needs to create
+      // a special chunk memory that hides reference to the allocator
+      // and call allocator's release function in the deleter.
+      return new T(std::forward<Args>(args)...);
+    }
+
+    static Object::FDeleter Deleter() {
+      return Deleter_;
+    }
+
+   private:
+    static void Deleter_(Object* ptr) {
+      auto storage_ptr = static_cast<StorageObj*>(ptr);
+      auto alloc = MemoryManager::Global()->GetAllocator(storage_ptr->buffer.ctx);
+      alloc->Free(storage_ptr->buffer);
+      delete ptr;
```

Comment:
Need to be very careful here. as ptr does not have virtual destructor, see the most correct way here https://github.com/dmlc/tvm/blob/master/include/tvm/runtime/memory.h#L116

You can also try to hide the ptr to the buffer if you like, in this particular case perhaps the best way is to subclass simple allocator and reuse it

Reply:
Okay I just deleted this code after thinking about the highlevel allocator, we can just use the destructor on the payload to release the buffer and not need our own allocator at all, we just need a custom allocation strategy for NDArray which is already in this PR>